editor/editor_file_system.cpp

```diff
@@ -3111,18 +3138,92 @@ Error EditorFileSystem::make_dir_recursive(const String &p_path, const String &p
 		efd->parent = parent;
 		efd->name = folder;
 		parent->subdirs.push_back(efd);
+		parent = efd;
+	}
 
-		if (first) {
-			parent->subdirs.sort_custom<DirectoryComparator>();
-			first = false;
+	_queue_refresh_filesystem();
+	return OK;
+}
+
+Error EditorFileSystem::copy_file(const String &p_from, const String &p_to) {
+	Ref<DirAccess> da = DirAccess::create(DirAccess::ACCESS_RESOURCES);
+	if (FileAccess::exists(p_from + ".import")) {
+		Error err = da->copy(p_from, p_to);
+		if (err != OK) {
+			return err;
+		}
+
+		// Remove uid from .import file to avoid conflict.
+		Ref<ConfigFile> cfg;
+		cfg.instantiate();
+		cfg->load(p_from + ".import");
+		cfg->erase_section_key("remap", "uid");
+		err = cfg->save(p_to + ".import");
+		if (err != OK) {
+			return err;
+		}
+
+		reimport_files_to_refresh.append(p_to);
+	} else if (ResourceLoader::get_resource_uid(p_from) == ResourceUID::INVALID_ID) {
+		// Files which do not use an uid can just be copied.
+		Error err = da->copy(p_from, p_to);
+		if (err != OK) {
+			return err;
+		}
+	} else {
+		// Load the resource and save it again in the new location (this generates a new UID).
+		Error err;
+		Ref<Resource> res = ResourceLoader::load(p_from, "", ResourceFormatLoader::CACHE_MODE_REUSE, &err);
+		if (err == OK && res.is_valid()) {
+			err = ResourceSaver::save(res, p_to, ResourceSaver::FLAG_COMPRESS);
+			if (err != OK) {
+				return err;
+			}
+		} else if (err != OK) {
+			// When loading files like text files the error is OK but the resource is still null.
+			// We can ignore such files.
+			return err;
 		}
-		parent = efd;
 	}
+	files_to_refresh.append(p_to);
 
-	emit_signal(SNAME("filesystem_changed"));
+	EditorFileSystemDirectory *parent = get_filesystem_path(p_to.get_base_dir());
+	ERR_FAIL_NULL_V(parent, ERR_FILE_NOT_FOUND);
+	folders_to_refresh.insert(parent);
+
+	EditorFileSystemDirectory::FileInfo *fi = memnew(EditorFileSystemDirectory::FileInfo);
+	fi->file = p_to.get_file();
+	parent->files.push_back(fi);
+
+	_queue_refresh_filesystem();
 	return OK;
 }
 
+void EditorFileSystem::copy_directory(const String &p_from, const String &p_to) {
```

Comment:
I think `copy_directory` should returns an `Error` and returns `OK` only if everything was moved correctly. First it's for consistency with other methods, and that way in `FileSystemDock` you could call `add_io_error` with a generic message when the copy did not end with a full success. Has you do with the file copy.

Reply:
I considered that, but when multiple files fail, not sure which error should be returned.